src/Addon.ts

```diff
@@ -0,0 +1,111 @@
+import { Request } from 'express';
+
+import { AuthError, AuthErrorCodes } from './AuthError';
+import { CustomTokenExtractor, Products, TokenPayload } from './types';
+import * as util from './util';
+
+export interface Credentials {
+  sharedSecret: string;
+}
+export type LoadCredentials = (clientKey: string) => Promise<Credentials | undefined>;
+export type SaveCredentials<RequestBody> = (
+  id: string,
+  body: RequestBody,
+  credentials?: Credentials
+) => Promise<Credentials>;
+
+export class Addon {
+  constructor(public product: Products, public baseUrl: string) {}
+
+  async install(
+    req: Request,
+    {
+      loadCredentials,
+      saveCredentials,
+    }: {
+      loadCredentials: LoadCredentials;
+      saveCredentials: SaveCredentials<Request['body']>;
+    }
+  ): Promise<{ credentials: Credentials; payload?: unknown }> {
+    const clientKey = util.extractClientKey(req);
+    const token = util.extractToken(req);
+    const credentials = await loadCredentials(clientKey);
+
+    if (token && util.extractIssuer(token) !== clientKey) {
+      throw new AuthError('Wrong issuer', AuthErrorCodes.WRONG_ISSUER);
+    }
+
+    // Create allowed if nothing was found by clientKey.
+    // Sometimes request signed (but we can't validate), sometimes not.
+    if (!credentials) {
+      const savedCredentials = await saveCredentials(clientKey, req.body);
+      return {
+        credentials: savedCredentials || req.body,
+      };
+    }
+
+    // Update allowed only if request was signed
+    if (credentials && token) {
+      const payload = util.validateToken(token, credentials.sharedSecret);
+
+      if (!payload.qsh && [Products.jira, Products.confluence].includes(this.product)) {
```

Comment:
Extracting temporary array to global constants would allow to prevent redundant array instantiation.